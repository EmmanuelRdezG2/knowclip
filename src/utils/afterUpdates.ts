import { defer, Observable } from 'rxjs'
import { mergeAll } from 'rxjs/operators'

export const afterUpdates = <T>(callback: () => Promise<Observable<T>>) =>
  defer(async () => {
    await null
    return await callback()
  }).pipe(mergeAll())
