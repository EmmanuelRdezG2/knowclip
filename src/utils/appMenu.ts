import electron from 'electron'

export function setAppMenuProjectSubmenuPermissions(projectOpened: boolean) {
  const menu = electron.remote.Menu.getApplicationMenu()
  const submenu = menu && menu.getMenuItemById('File').submenu
  if (!submenu) return

  submenu.getMenuItemById('Save project').enabled = projectOpened
  submenu.getMenuItemById('Close project').enabled = projectOpened

  submenu.getMenuItemById('Open project').enabled = !projectOpened
}
