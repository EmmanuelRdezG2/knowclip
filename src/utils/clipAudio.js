import ffmpeg, { toTimestamp } from '../utils/ffmpeg'

const clipAudio = (sourceFilePath, startTime, endTime, outputFilename) => {
  return new Promise((res, rej) => {
    ffmpeg(sourceFilePath)
      // .audioCodec('copy') // later, do this and change hardcoded '.mp3' for audio-only input
      .seekInput(toTimestamp(startTime))
      .inputOptions('-to ' + toTimestamp(endTime))
      .outputOptions('-vn')
      .output(outputFilename)
      .on(
        'end',
        //listener must be a function, so to return the callback wrapping it inside a function
        function() {
          res(outputFilename)
        }
      )
      .on('error', err => {
        console.error(err)
        rej(err)
      })
      .run()
  })
}
export default clipAudio
