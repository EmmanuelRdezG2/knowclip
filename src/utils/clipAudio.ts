import ffmpeg, { toTimestamp } from '../utils/ffmpeg'
import { existsSync } from 'fs-extra'

import { Writable } from 'stream'

export const clipAudioStream = (
  sourceFilePath: string,
  startTime: number,
  endTime: number,
  outputFilename: string
): { done: Promise<string>; stream: Writable } => {
  const mp3 = ffmpeg(sourceFilePath)
    // .audioCodec('copy') // TODO: do this and change hardcoded '.mp3' for audio-only input
    .seekInput(toTimestamp(startTime))
    .inputOptions('-to ' + toTimestamp(endTime))
    .outputOptions('-vn')
    .outputFormat('mp3')
  return {
    done: new Promise((res, rej) => {
      mp3
        .on('end', () => {
          console.log('done!', outputFilename)
          res(outputFilename)
        })
        .on('error', err => {
          console.error(err)
          rej(err)
        })
    }),
    stream: mp3.pipe(),
  }
}

const clipAudio = async (
  sourceFilePath: string,
  startTime: number,
  endTime: number,
  outputFilename: string
): Promise<string> => {
  const result = await new Promise((res, rej) => {
    ffmpeg(sourceFilePath)
      // .audioCodec('copy') // TODO: do this and change hardcoded '.mp3' for audio-only input
      .seekInput(toTimestamp(startTime))
      .inputOptions('-to ' + toTimestamp(endTime))
      .outputOptions('-vn')
      .output(outputFilename)
      .on('end', () => {
        res(outputFilename)
      })
      .on('error', err => {
        console.error(err)
        rej(err)
      })
      .run()
  })

  if (!result || !existsSync(outputFilename)) {
    console.log({ outputFilename, sourceFilePath, startTime, endTime })
    throw new Error(`Problem clipping audio from ${sourceFilePath}`)
  }

  return outputFilename
}
export default clipAudio
