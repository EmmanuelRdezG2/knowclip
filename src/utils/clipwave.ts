export const CLIPWAVE_ID = 'waveform' as const
