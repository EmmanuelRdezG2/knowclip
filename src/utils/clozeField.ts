type RawClozeRange = {
  startIndex: number
  fullMatchText: string
  tagContents: string
}

export const unescapeClozeFields = <F extends FlashcardFields>(
  fieldsFromJson?: Partial<F>
): { fields: Partial<F>; cloze: ClozeDeletion[] } => {
  console.log('UNESCAPING!!')
  const fields = {} as F
  if (!fieldsFromJson) return { fields, cloze: [] }
  const cloze: ClozeDeletion[] = []
  for (const fn in fieldsFromJson) {
    const fieldName: keyof F = fn
    const value: string = fieldsFromJson[fieldName] as any
    if (fieldName === 'transcription') {
      const { text, clozeDeletions } = unescapeClozeField(value)
      fields[fieldName] = (text as unknown) as F[keyof F]
      cloze.push(...clozeDeletions)
    } else {
      fields[fieldName] = (value as unknown) as F[keyof F]
    }
  }

  return { fields, cloze }
}

const clozeTagsRegex = new RegExp(`<c(10|[1-9])>(.*?)</c>`, 'g')

export function unescapeClozeField(text: string) {
  const matches = [...text.matchAll(clozeTagsRegex)]

  let result = ''
  const rawRangeGroups = matches
    .reduce(
      (deletions, match) => {
        const [fullMatchText, clozeNumberRaw, tagContents] = match
        const index = typeof match.index === 'number' ? match.index : -1
        const clozeNumber = +clozeNumberRaw

        deletions[clozeNumber - 1] = deletions[clozeNumber - 1] || []
        deletions[clozeNumber - 1].push({
          startIndex: index,
          fullMatchText,
          tagContents,
        })
        return deletions
      },
      [] as RawClozeRange[][]
    )
    .filter(r => r && r.length)

  if (!rawRangeGroups.length) return { text, clozeDeletions: [] }

  const deletions: ClozeDeletion[] = rawRangeGroups.map(
    (rawRanges, rangeGroupIndex) => {
      const ranges: ClozeRange[] = []

      const firstRange = rawRanges[0]
      if (rangeGroupIndex === 0 && firstRange.startIndex > 0) {
        result += text.slice(0, firstRange.startIndex)
      }

      let i = 0
      for (const rawRange of rawRanges) {
        ranges.push({
          start: result.length,
          end: result.length + rawRange.tagContents.length,
        })
        result += rawRange.tagContents

        const nextRange = rawRanges[i + 1]
        const subsequentGapEnd = nextRange ? nextRange.startIndex : text.length
        result += text.slice(
          rawRange.startIndex + rawRange.fullMatchText.length,
          subsequentGapEnd
        )

        i++
      }

      return { ranges }
    }
  )

  return { text: result, clozeDeletions: deletions }
}
