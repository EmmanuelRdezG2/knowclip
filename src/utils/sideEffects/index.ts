import sideEffects from './module'
import spectronMocks from '../../test/spectronMocks'

const {
  mocked,
  mockFunctions: mockSideEffects,
  logMocks: logSideEffectsMocks,
} = spectronMocks('side-effect', sideEffects)

export { mockSideEffects, logSideEffectsMocks }

const { uuid, nowUtcTimestamp } =
  process.env.REACT_APP_CHROMEDRIVER || process.env.NODE_ENV === 'test'
    ? mocked
    : sideEffects

export { uuid, nowUtcTimestamp }
