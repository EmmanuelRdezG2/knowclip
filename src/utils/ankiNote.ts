import { join, basename } from 'path'
import { promises } from 'fs'
import clipAudio from '../utils/clipAudio'
import { getVideoStill } from '../utils/getVideoStill'
const { readFile } = promises

export type AnkiNoteMedia = {
  soundData: {
    data: () => Promise<Buffer>
    fileName: string
    filePath: string
  }
  imageData: {
    data: () => Promise<Buffer>
    fileName: string
    filePath: string
  } | null
}

export function getClozeFields(fields: string[], clozeDeletions: string) {
  return [clozeDeletions, ...fields.slice(1)]
}

export async function processNoteMedia(
  clipSpecs: ClipSpecs,
  directory: string
): AsyncResult<AnkiNoteMedia> {
  const {
    outputFilename,
    sourceFilePath,
    startTime,
    endTime,
    flashcardSpecs: { image },
  } = clipSpecs

  const clipOutputFilePath = join(directory, outputFilename)
  const clipAudioResult = await clipAudio(
    sourceFilePath,
    startTime,
    endTime,
    clipOutputFilePath
  )
  if (clipAudioResult.errors)
    return {
      errors: [
        `Could not make clip from ${sourceFilePath}`,
        ...clipAudioResult.errors,
      ],
    }
  const soundData = {
    data: () => readFile(clipOutputFilePath),
    fileName: outputFilename,
    filePath: clipOutputFilePath,
  }

  const imageResult = image
    ? await getVideoStill(image.id, sourceFilePath, image.seconds)
    : null
  if (imageResult && imageResult.errors)
    return {
      errors: [
        `Could not make still from ${sourceFilePath}`,
        ...imageResult.errors,
      ],
    }

  const imageData = imageResult
    ? {
        data: () => readFile(imageResult.value),
        fileName: basename(imageResult.value),
        filePath: imageResult.value,
      }
    : null

  return {
    value: {
      soundData,
      imageData,
    },
  }
}
