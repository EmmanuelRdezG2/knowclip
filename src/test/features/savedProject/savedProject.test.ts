import { Application } from 'spectron'
import {
  startApp,
  stopApp,
  TestSetup,
  TMP_DIRECTORY,
  ASSETS_DIRECTORY,
  GENERATED_ASSETS_DIRECTORY,
} from '../../spectronApp'
import { mockSideEffects } from '../../../utils/sideEffects'
import openSavedProject from './openSavedProject'
import { join } from 'path'
import makeSomeFlashcards from './makeSomeFlashcards'
import saveAndCloseProject from './saveAndCloseProject'

jest.setTimeout(60000)

describe('opening and saving a previously saved project', () => {
  let context: { app: Application | null } = { app: null }
  let setup: TestSetup

  beforeAll(async () => {
    setup = await startApp(context, 'savedProject', persistedState)

    await mockSideEffects(setup.app, sideEffectsMocks)
  })

  test('opens a previously saved project', () => openSavedProject(setup))
  test('make some flashcards', () => makeSomeFlashcards(setup))
  test('saves and closes project', () => saveAndCloseProject(setup))

  afterAll(async () => {
    await stopApp(context)
  })
})

const sideEffectsMocks = {
  uuid: [
    'ece1aa0c-2377-47db-b5d1-47d63e918ca7',
    'cab0801b-3d8d-4cf8-a8d0-575beed8d645',
    'fcf1bb3f-a5a7-419d-9c2c-2d93252f8b71',
  ],
  nowUtcTimestamp: ['2020-01-22T07:50:05Z', '2020-01-22T07:50:10Z'],
}

const persistedState: Partial<AppState> = {
  files: {
    ProjectFile: {
      'e36ca6de-4893-49f5-bca1-4410ace25d46': {
        type: 'ProjectFile',
        id: 'e36ca6de-4893-49f5-bca1-4410ace25d46',
        name: 'My cool saved project',
        noteType: 'Transliteration',
        mediaFileIds: [
          '535935a3-7d72-4238-87ab-1b7a413c1f71',
          'fa462524-bfeb-4434-a506-07858cc97151',
        ],
        error: null,
        lastOpened: '2020-01-20T11:43:27Z',
        lastSaved: '2020-01-20T11:42:45Z',
      },
    },
    MediaFile: {
      '535935a3-7d72-4238-87ab-1b7a413c1f71': {
        id: '535935a3-7d72-4238-87ab-1b7a413c1f71',
        type: 'MediaFile',
        parentId: 'e36ca6de-4893-49f5-bca1-4410ace25d46',
        subtitles: [
          {
            type: 'EmbeddedSubtitlesTrack',
            id: '37bd5e91-89c5-491b-9308-533c7be7338a',
            streamIndex: 2,
          },
          {
            type: 'ExternalSubtitlesTrack',
            id: '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439',
          },
        ],
        flashcardFieldsToSubtitlesTracks: {},
        name: 'polar_bear_cafe.mp4',
        durationSeconds: 87.062,
        format: 'mov,mp4,m4a,3gp,3g2,mj2',
        isVideo: true,
        subtitlesTracksStreamIndexes: [2],
      },
      'fa462524-bfeb-4434-a506-07858cc97151': {
        id: 'fa462524-bfeb-4434-a506-07858cc97151',
        type: 'MediaFile',
        parentId: 'e36ca6de-4893-49f5-bca1-4410ace25d46',
        subtitles: [],
        flashcardFieldsToSubtitlesTracks: {},
        name: 'piggeldy_cat.mp4',
        durationSeconds: 58.24,
        format: 'mov,mp4,m4a,3gp,3g2,mj2',
        isVideo: true,
        subtitlesTracksStreamIndexes: [],
      },
    },
    ExternalSubtitlesFile: {
      '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439': {
        type: 'ExternalSubtitlesFile',
        parentId: '535935a3-7d72-4238-87ab-1b7a413c1f71',
        id: '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439',
        name: 'pbc_jp.ass',
      },
    },
    VttConvertedSubtitlesFile: {
      '37bd5e91-89c5-491b-9308-533c7be7338a': {
        type: 'VttConvertedSubtitlesFile',
        parentId: '535935a3-7d72-4238-87ab-1b7a413c1f71',
        id: '37bd5e91-89c5-491b-9308-533c7be7338a',
        streamIndex: 2,
        parentType: 'MediaFile',
      },
      '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439': {
        type: 'VttConvertedSubtitlesFile',
        id: '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439',
        parentId: '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439',
        parentType: 'ExternalSubtitlesFile',
      },
    },
    WaveformPng: {
      '535935a3-7d72-4238-87ab-1b7a413c1f71': {
        type: 'WaveformPng',
        parentId: '535935a3-7d72-4238-87ab-1b7a413c1f71',
        id: '535935a3-7d72-4238-87ab-1b7a413c1f71',
      },
      'fa462524-bfeb-4434-a506-07858cc97151': {
        type: 'WaveformPng',
        parentId: 'fa462524-bfeb-4434-a506-07858cc97151',
        id: 'fa462524-bfeb-4434-a506-07858cc97151',
      },
    },
    ConstantBitrateMp3: {},
  },
  fileAvailabilities: {
    ProjectFile: {
      'e36ca6de-4893-49f5-bca1-4410ace25d46': {
        filePath: join(TMP_DIRECTORY, 'my_previously_saved_project.afca'),
        status: 'CURRENTLY_LOADED',
        id: 'e36ca6de-4893-49f5-bca1-4410ace25d46',
      },
    },
    MediaFile: {
      '535935a3-7d72-4238-87ab-1b7a413c1f71': {
        filePath: join(ASSETS_DIRECTORY, 'polar_bear_cafe.mp4'),
        status: 'CURRENTLY_LOADED',
        id: '535935a3-7d72-4238-87ab-1b7a413c1f71',
      },
      'fa462524-bfeb-4434-a506-07858cc97151': {
        filePath: join(ASSETS_DIRECTORY, 'piggeldy_cat.mp4'),
        status: 'CURRENTLY_LOADED',
        id: 'fa462524-bfeb-4434-a506-07858cc97151',
      },
    },
    ExternalSubtitlesFile: {
      '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439': {
        filePath: join(ASSETS_DIRECTORY, 'pbc_jp.ass'),
        status: 'CURRENTLY_LOADED',
        id: '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439',
      },
    },
    VttConvertedSubtitlesFile: {
      '37bd5e91-89c5-491b-9308-533c7be7338a': {
        filePath: join(
          GENERATED_ASSETS_DIRECTORY,
          '95ecfad8a8bf42d562d0e751aa9f8448.vtt'
        ),
        status: 'CURRENTLY_LOADED',
        id: '37bd5e91-89c5-491b-9308-533c7be7338a',
      },
      '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439': {
        filePath: join(
          GENERATED_ASSETS_DIRECTORY,
          '3e59a4e5742e827fdb30511fc40f7e82.vtt'
        ),
        status: 'CURRENTLY_LOADED',
        id: '2f7a98c0-1bf8-44cf-9b0b-4af64fb86439',
      },
    },
    WaveformPng: {
      '535935a3-7d72-4238-87ab-1b7a413c1f71': {
        filePath: join(
          GENERATED_ASSETS_DIRECTORY,
          '833a0a72cd50d38e26f320c7924f4da6.png'
        ),
        status: 'CURRENTLY_LOADED',
        id: '535935a3-7d72-4238-87ab-1b7a413c1f71',
      },
      'fa462524-bfeb-4434-a506-07858cc97151': {
        filePath: join(
          GENERATED_ASSETS_DIRECTORY,
          'f492c13cdbdf0c59adf3a72b2cb74e04.png'
        ),
        status: 'CURRENTLY_LOADED',
        id: 'fa462524-bfeb-4434-a506-07858cc97151',
      },
    },
    ConstantBitrateMp3: {},
  },
}
