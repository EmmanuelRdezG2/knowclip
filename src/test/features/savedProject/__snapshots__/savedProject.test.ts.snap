// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`opening and saving a previously saved project resulting project file matches snapshot 1`] = `
Object {
  "value": Object {
    "media": Array [
      Object {
        "clips": Array [
          Object {
            "end": "00:12.5",
            "fields": Object {
              "meaning": "Relaxing while eating bamboo grass is the best",
              "pronunciation": "sasa-o tabe-nágara nonbíri-suru-no-wa saikoo-da-naa",
              "transcription": "笹を食べながらのんびりするのは最高だなぁ",
            },
            "id": "632a6cff-7fd7-4d0f-b657-0b9636204261",
            "image": Object {
              "type": "VideoStill",
            },
            "start": "00:08.04",
            "tags": Array [
              "pbc",
            ],
          },
          Object {
            "end": "00:17.76",
            "fields": Object {
              "meaning": "This kid, lazing about again so early",
              "notes": "\\"Goro-goro\\" is the sound of something big rolling around.",
              "pronunciation": "mata kono ko-wa hirumá-kara góro-goro shite",
              "transcription": "ちょっと  またこの子は昼間からゴロゴロして",
            },
            "id": "a272b282-4d89-4af0-9d17-baa807793f1f",
            "start": "00:15.12",
            "tags": Array [
              "pbc",
            ],
          },
          Object {
            "end": "01:21.16",
            "fields": Object {
              "transcription": "えーと  働かないでお金がもらえる仕事が載っているのはどれですか",
            },
            "id": "6fce08e8-9312-4930-9601-c14b470e57ab",
            "image": Object {
              "type": "VideoStill",
            },
            "start": "01:15.2",
            "tags": Array [
              "pbc",
            ],
          },
        ],
        "duration": "01:27.062",
        "flashcardFieldsToSubtitlesTracks": Object {
          "transcription": "b3347257-97c4-4c9a-8eae-9bf88a8edc23",
        },
        "format": "mov,mp4,m4a,3gp,3g2,mj2",
        "id": "10fd95ee-ba1a-424f-95fb-66a28c91faf1",
        "name": "polar_bear_cafe.mp4",
        "subtitles": Array [
          Object {
            "id": "d680807d-3a5f-4a25-967f-71871b7d5057",
            "streamIndex": 2,
            "type": "Embedded",
          },
          Object {
            "id": "b3347257-97c4-4c9a-8eae-9bf88a8edc23",
            "name": "pbc_jp.ass",
            "type": "External",
          },
        ],
      },
      Object {
        "clips": Array [
          Object {
            "end": "00:08.76",
            "fields": Object {
              "meaning": "I am not a cat, said Frederick angrily",
              "transcription": "Ich bin keine Katze, sagte Frederick böse",
            },
            "id": "136042f2-00bc-4d9b-a3a7-3da1b7868c78",
            "image": Object {
              "type": "VideoStill",
            },
            "start": "00:07.02",
            "tags": Array [
              "piggeldy_cat.mp4",
            ],
          },
          Object {
            "end": "00:20.0",
            "fields": Object {
              "meaning": "I didn't say that",
              "transcription": "Das hab' ich nicht gesagt",
            },
            "id": "1e62d23d-3f3b-4785-b8ba-5d61dddc78ed",
            "image": Object {
              "type": "VideoStill",
            },
            "start": "00:18.42",
            "tags": Array [
              "piggeldy_cat.mp4",
            ],
          },
        ],
        "duration": "00:58.24",
        "format": "mov,mp4,m4a,3gp,3g2,mj2",
        "id": "bb2bd7fe-4ead-48aa-bad5-88df7a1b5e77",
        "name": "piggeldy_cat.mp4",
      },
    ],
    "project": Object {
      "id": "91bfd159-155c-4b61-bdd5-d71e2e944773",
      "name": "My cool saved project",
      "noteType": "Transliteration",
      "timestamp": "2020-02-02T15:31:43Z",
    },
  },
}
`;
